scripts/o3de/o3de/cmake.py

```diff
@@ -27,75 +29,55 @@ def add_gem_dependency(cmake_file: pathlib.Path,
-    end_marker_line_index = None
+    gem_name_with_version_specifier = gem_name
+    gem_name, _ = utils.get_object_name_and_optional_version_specifier(gem_name)
+
+    pre_gem_names = ''
+    post_gem_names = ''
+    all_gem_names = '' 
+    indent = 4
+
     with cmake_file.open('r') as s:
-        in_gem_list = False
-        line_index = 0
-        for line in s:
-            parsed_line = line.strip()
-            if parsed_line.startswith(enable_gem_start_marker):
-                # Skip pass the 'set(ENABLED_GEMS' marker just in case their are gems declared on the same line
-                parsed_line = parsed_line[len(enable_gem_start_marker):]
-                # Set the flag to indicate that we are in the ENABLED_GEMS variable
-                in_gem_list = True
-                start_marker_line_index = line_index
+        file_contents = s.read()
+        if file_contents:
+            # regex to isolate pre, gem names and post 
+            regex_all_gem_names = re.compile('(?P<pre>[\s\S]*set\s*\(\s*ENABLED_GEMS\s*)?(?P<gem_names>[^\)]+)(?P<post>\)[\s\S]*)?')
+            all_gem_names_matches = regex_all_gem_names.search(file_contents)
+            if not all_gem_names_matches:
+                logger.error(f'{cmake_file} is not formatted correctly and cannot be modified.')
+                return 1
 
-            if in_gem_list:
-                # Since we are inside the ENABLED_GEMS variable determine if the line has the end_marker of ')'
-                if parsed_line.endswith(enable_gem_end_marker):
-                    # Strip away the line end marker
-                    parsed_line = parsed_line[:-len(enable_gem_end_marker)]
-                    # Set the flag to indicate that we are no longer in the ENABLED_GEMS variable after this line
-                    in_gem_list = False
-                    end_marker_line_index = line_index
+            pre_gem_names = all_gem_names_matches.group('pre')
+            post_gem_names = all_gem_names_matches.group('post')
+            if pre_gem_names and not post_gem_names:
+                logger.error(f'{cmake_file} is missing a closing parenthesis or is not formatted correctly.')
+                return 1
 
-                # Split the rest of the line on whitespace just in case there are multiple gems in a line
-                gem_name_list = map(lambda gem_name: gem_name.strip('"'), parsed_line.split())
-                if gem_name in gem_name_list:
-                    logger.info(f'{gem_name} is already enabled in file {str(cmake_file)}.')
-                    return 0
+            # regex to replace every version of the gem name 
+            regex_replace_gem_name = re.compile(f'(?:^|\s+){gem_name}[=><~]*[\S]*')
+            all_gem_names = regex_replace_gem_name.sub('', all_gem_names_matches.group('gem_names'))
 
-            t_data.append(line)
-            line_index += 1
+            all_gem_names += f'\n{" "  * indent}{gem_name_with_version_specifier}'
```

Comment:
The reason regular expressions weren't used is because it is not powerful enough to parse all the obscure cases like having multiple gems on the same with version specifiers in them or the fact that the ENABLED_GEMS might not be terminated correctly with a trailing `)` causing further corruption of the enable_gems.cmake when writing back to it.

Reply:
Do the unit tests cover these edge cases? If so, then they are currently passing with these regex.

Reply:
The Unit test does cover the complexity for regular gem names.
Test would have to be added to make sure specifying gem dependencies on the same line would work.
i.e
```cmake
set(ENABLED_GEMS
    gem1=1.0.0 gem2=2.0.0 gem3 gem4=1.0.1)
```

Reply:
I do have this test:
```
            # when multiple versions of the same gem are found expect all are replaced with the new entry
            pytest.param("""
                        # Comment
                        set(ENABLED_GEMS
                            foo bar
                            TestGem==1.0.0 TestGem TestGem>=2.0.0
                            baz)
                    """, 'TestGem', set(['foo', 'bar', 'baz', 'TestGem']), 0),
```

Reply:
That should work.
Can you add a test make sure the `enable_gems.cmake` file isn't modified if `set(ENABLED_GEMS)` variable is invalid such as it is missing an ending `)`

Reply:
As discussed, I've deprecated `enabled_gems.cmake` and the functions that add gems to it